From 37856dac7c66137662833b2ef73a97a8537d0373 Mon Sep 17 00:00:00 2001
From: Sai Rupesh Chevuru <chevuru@qti.qualcomm.com>
Date: Thu, 7 Aug 2025 18:06:36 +0530
Subject: [PATCH] QSDK:DP:wifi:mac80211: skip fast path offload when skb mark
 is invalid

This patch addresses a stability issue in fast forwarding scenarios
where the skb->mark field may not be properly populated by forwarding engines.
In MLO VAP configurations, delivering such packets directly to the driver
can cause enqueuing to wrong primary and leading to a packets drop.

To mitigate this, a validity check for the skb->mark field has been
added before offloading the packet directly to the driver.

Patch-dependency:
0126-mac80211-Deliver-the-frame-to-driver-tx-ops-directly.patch

Signed-off-by: Sai Rupesh Chevuru <chevuru@qti.qualcomm.com>
---
 net/mac80211/tx.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index 8c9b28d..7d2103b 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -5032,6 +5032,10 @@ netdev_tx_t ieee80211_subif_start_xmit_8023(struct sk_buff *skb,
 #endif
 	orig_sdata = sdata;
 	if (likely(skb->fast_xmit && perf_mode)) {
+		if (unlikely(((skb->mark & 0xFF800000) != 0xAA800000) &&
+			       ieee80211_vif_is_mld(&sdata->vif)))
+			goto regular_xmit;
+
 		if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN)
 			sdata = container_of(sdata->bss,
 					     struct ieee80211_sub_if_data, u.ap);
@@ -5047,6 +5051,7 @@ netdev_tx_t ieee80211_subif_start_xmit_8023(struct sk_buff *skb,
 		return NETDEV_TX_OK;
 	}
 
+regular_xmit:
 	info->control.vif = &sdata->vif;
 #else
 #ifdef CPTCFG_MAC80211_ATHMEMDEBUG
-- 
2.34.1

